File: android/src/main/kotlin/io/sentient/android/settings/audio/AudioViewModel.kt
// ---------------------------------------------------------------------------
// AudioViewModel — Audio settings page. FAST save: a full-profile PUT whose only
// diff is `audio`, so ApplyProfileChangeUseCase takes the no-restart fast path and
// pushes a live WS preference patch. Holds `original` (server truth) + `draft`
// (local edits); dirty is derived. Thin: loads via profileRepository, saves via the
// one FSM usecase, folds its ApplyState flow into flat progress fields.
// ---------------------------------------------------------------------------
package io.sentient.android.settings.audio

import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import io.sentient.mobiledata.di.SettingsComponent
import io.sentient.mobiledata.result.SentientResult
import io.sentient.mobiledata.usecase.settings.ApplyState
import io.sentient.mobiledata.usecase.settings.ProfileMutation
import io.sentient.mobilesdk.log.createLogger
import io.sentient.mobilesdk.settings.ProfileV1
import io.sentient.mobilesdk.settings.toPutBody
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.flow.update
import kotlinx.coroutines.launch

/** Audio page state. `original` is the last server truth; `draft` carries local edits. */
data class AudioUiState(
    val loading: Boolean = true,
    val loadError: String? = null,
    val original: ProfileV1? = null,
    val draft: ProfileV1? = null,
    val saving: Boolean = false,
    val alreadyApplying: Boolean = false,
    val applyError: String? = null,
) {
    /** Dirty iff the audio section of the draft diverges from the server truth. */
    val dirty: Boolean
        get() = original != null && draft != null && draft.audio != original.audio
}

class AudioViewModel(private val component: SettingsComponent) : ViewModel() {
    private val log = createLogger("android", "settings", "audio-vm")

    private val _ui = MutableStateFlow(AudioUiState())
    val ui: StateFlow<AudioUiState> = _ui.asStateFlow()

    init {
        load()
    }

    private fun load() {
        viewModelScope.launch {
            _ui.update { it.copy(loading = true, loadError = null) }
            when (val r = component.profileRepository.getProfile()) {
                is SentientResult.Success ->
                    _ui.update { it.copy(loading = false, original = r.data, draft = r.data) }
                is SentientResult.Failure -> {
                    log.warn("load.failed", mapOf("kind" to r.error.kind))
                    _ui.update { it.copy(loading = false, loadError = r.error.userMessage) }
                }
                is SentientResult.Loading -> Unit
            }
        }
    }

    fun setTts(on: Boolean) = editDraft { it.copy(audio = it.audio.copy(ttsEnabled = on)) }

    fun setChannel(channel: String) = editDraft { it.copy(audio = it.audio.copy(channel = channel)) }

    private inline fun editDraft(transform: (ProfileV1) -> ProfileV1) {
        _ui.update { s ->
            val draft = s.draft ?: return@update s
            s.copy(draft = transform(draft), alreadyApplying = false, applyError = null)
        }
    }

    fun save() {
        val state = _ui.value
        val previous = state.original ?: return
        val next = state.draft ?: return
        if (!state.dirty || state.saving) return
        log.info("save", mapOf("tts" to next.audio.ttsEnabled, "channel" to next.audio.channel))
        viewModelScope.launch {
            component.applyProfileChange(ProfileMutation.PutProfile(previous, next.toPutBody())).collect { st ->
                _ui.update { it.foldApply(st) }
                if (st is ApplyState.Ready) refetch()
            }
        }
    }

    private suspend fun refetch() {
        when (val r = component.profileRepository.getProfile()) {
            is SentientResult.Success -> _ui.update { it.copy(original = r.data, draft = r.data) }
            is SentientResult.Failure -> log.warn("refetch.failed", mapOf("kind" to r.error.kind))
            is SentientResult.Loading -> Unit
        }
    }
}

/** Fold one FSM transition into flat progress fields (fast path never emits Restarting). */
internal fun AudioUiState.foldApply(state: ApplyState): AudioUiState = when (state) {
    ApplyState.Saving -> copy(saving = true, alreadyApplying = false, applyError = null)
    ApplyState.Restarting -> copy(saving = true)
    is ApplyState.Ready -> copy(saving = false, alreadyApplying = false, applyError = null)
    ApplyState.AlreadyApplying -> copy(saving = false, alreadyApplying = true)
    is ApplyState.Failed -> copy(saving = false, applyError = state.error.userMessage)
    ApplyState.Idle -> this
}

File: ios/App/Settings/Audio/AudioViewModel.swift
// ---------------------------------------------------------------------------
// AudioViewModel — the Audio settings page state holder. FAST save: a profile
// PUT whose only diff is `audio` takes ApplyProfileChangeUseCase's audio-only
// fast path (PUT + live WS preference patch, NO Hermes restart). So this page's
// Save never shows the "restarting…" copy — Ready lands immediately.
//
// Holds the loaded ProfileV1 as `original` (the mutation's `previous`) plus the
// two Swift-native draft primitives the controls bind to. Dirty is derived from
// those primitives (KMP data classes aren't Swift-Equatable, so we never compare
// ProfileV1 directly). Save rebuilds `next` and collects the FSM flow.
// ---------------------------------------------------------------------------
import Foundation
import MobileData

@MainActor
@Observable
final class AudioViewModel {
    /// Initial-load phase for the page body.
    enum Phase: Equatable {
        case loading
        case ready
        case failed(String)
    }

    /// Save FSM projection. Audio is fast so `.restarting` never occurs, but the
    /// case is kept so the projection matches the shared ApplyState alphabet.
    enum Save: Equatable {
        case idle
        case saving
        case restarting
        case alreadyApplying
        case failed(String)
    }

    private(set) var phase: Phase = .loading
    private(set) var save: Save = .idle

    /// Draft primitives bound to the controls.
    var ttsEnabled = false
    var channel = ProfileEnums.shared.audioChannels.first ?? "voice"

    private var original: ProfileV1?
    private let settings: SettingsComponent
    private let log = AppLog("settings", "audio-vm")

    init(settings: SettingsComponent) {
        self.settings = settings
    }

    /// True once the draft differs from the loaded profile's audio section.
    var isDirty: Bool {
        guard let o = original else { return false }
        return ttsEnabled != o.audio.ttsEnabled || channel != o.audio.channel
    }

    /// True while a save is in flight (Save is disabled, a banner shows).
    var isApplying: Bool { save == .saving || save == .restarting }

    /// Load the profile and seed original + draft. Folds the envelope exhaustively.
    func load() async {
        log.info("load")
        do {
            let result = try await settings.profileRepository.getProfile()
            switch onEnum(of: result) {
            case .success(let s):
                apply(s.data)
                phase = .ready
                log.info("load.ready tts=\(ttsEnabled) channel=\(channel)")
            case .failure(let f):
                phase = .failed(f.error.userMessage)
                log.warn("load.failed kind=\(f.error.kind)")
            case .loading:
                phase = .loading
            }
        } catch is CancellationError {
            // View gone — not a failure.
        } catch {
            phase = .failed("Couldn't load audio settings.")
            log.warn("load.threw")
        }
    }

    /// Fast save: PUT profile → live audio patch (no restart) → refetch truth.
    func save() async {
        guard let o = original, isDirty else { return }
        log.info("save.start tts=\(ttsEnabled) channel=\(channel)")
        let mutation = ProfileMutationPutProfile(previous: o, next: nextProfile(from: o))
        for await state in settings.applyProfileChange.invoke(mutation: mutation) {
            switch onEnum(of: state) {
            case .idle: break
            case .saving: save = .saving
            case .restarting: save = .restarting
            case .ready:
                save = .idle
                log.info("save.ready")
                await load()
            case .alreadyApplying:
                save = .alreadyApplying
                log.warn("save.already-applying")
            case .failed(let f):
                save = .failed(f.error.userMessage)
                log.warn("save.failed")
            }
        }
    }

    private func apply(_ profile: ProfileV1) {
        original = profile
        ttsEnabled = profile.audio.ttsEnabled
        channel = profile.audio.channel
    }

    private func nextProfile(from o: ProfileV1) -> ProfileV1 {
        ProfileV1(
            schemaVersion: o.schemaVersion,
            userId: o.userId,
            model: o.model,
            voice: o.voice,
            audio: ProfileAudio(ttsEnabled: ttsEnabled, channel: channel),
            persona: o.persona,
            tools: o.tools,
            compression: o.compression,
            advanced: o.advanced,
            devices: o.devices
        )
    }
}
